### DISCUSSION TITLE
Pull-up registers not recognised on 828?

### DISCUSSION THREAD

**Discussion by fxframes** [SpenceKonde/ATTinyCore#745] [Support / Q & A]
Hi,

When trying something like PUEB = (1<<PUEB7); to enable a pull-up resistor on an ATTiny 828 I get a compile error.

error: 'PUEB7' was not declared in this scope
   PUEB |= (1<<PUEB7);

I double-checked that that is indeed the correct register name and bit names (from the 828 data sheet):

PUEB – Port B Pull-Up Enable Control Register
Bit 76543210
0x07 (0x27) PUEB7 PUEB6 PUEB5 PUEB4 PUEB3 PUEB2 PUEB1 PUEB0 PUEB

Please advise, thanks!

**Comment by hmeijdam** [comment]
PORTA_PIN2CTRL = PORT_PULLUPEN_bm; // enable PortA Pin2 pullup

etc.

**Reply by SpenceKonde** [reply]
He's on ATTinyCore, using an ATtiny828, that's a classic AVR, not a modern AVR. But it's one of the wacky three classic tinies with PUE registers.

**Comment by hmeijdam** [comment]
Oops. My bad. Sorry--Sent from my Android phone with GMX Mail. Please excuse my brevity.On 30-01-2023, 15:54 "Spence Konde (aka Dr. Azzy)" ***@***.***> wrote:

   
  He's on ATTinyCore, using an ATtiny828, that's a classic AVR, not a modern AVR. But it's one of the wacky three classic tinies with PUE registers.

**Comment by SpenceKonde** [comment]
Yes, I agree. But Atmel/Microchip has not seen fit to define those bits by name.... 

Personally, I don't think I have *ever* seen someone on a classic AVR using the names of the bits to refer to the pins. PINA0-7, PORTA0-7, DDA0-7 and the same bits for all the other ports are just macros #defined to be the numbers 0 through 7.  People generally just use the numeric values, 

```
PUEB |= 1 << 7;
````
Or, I guess
```
PUEB |= 1<< PINB7 
```
Works too. 

As does the arduino pinMode() function. 

For whatever reason, they just didn't put PUExn bits into the datasheet by name. (my guess: Back when the part was released, management yanked on the engineering team's leash the instant it hit the ground and dragged them to the modern AVR development team. Before they'd really polished the headers, and despite the fact that the chip's planned headline feature was airbrushed out of the datasheet (It's very clear that it was supposed to have the same fancy ADC that the 841 got, only with more inputs - look at the register layout for the ADC, vs the register layout for the ADC on any other classic AVRs. Those two parts, from the same vintage are the only ones with that layout. Except... only the 841 got a working differential ADC. Based on the Atmel/Microchip development cycle, and history what I think happened is: 2000's: Atmel bets big on xMega, and let's the classic AVRs start to become a backwater. 
2010ish : Atmel finally realizes that the battle is over, and xMega lost (who could have imagined? The parts were incomprehensibly complex like an ARM, but the instruction set was still 8 bits so they lacked computational power, they had the lower maximum operating voltage of typical ARMs, and they cost an arm and a leg. They basically combined the disadvantages of AVR and ARM, instead of the advantages, and Atmel sent them into battle against the ARM processors - they had less of a chance than a single snowball would have of freezing over all hell!) Three (4 if you count the 441 as distinct) tinyAVR testbed parts start the design process to serve as trial balloons for new ideas, because that's what they did and that's what they have done in the years since (the tinyAVR 1-series, halfway through releasing it, when they got to the bigger memory versions, they stuffed some extra features in that they wanted to try out for the Dx-series parts. And the 2-series tinyAVRs are the trial balloon for the EA-series ADC)
2013: They've been playing with the 841. They've learned some things and realized they need to think bigger, some engineers are already working on what would become the modern AVRs., and so on. Die rev B of the silicon for the 828 comes back from the fab, the datasheet is all written, they're ready to ship - and they desperately want to get all the designers working on the modern AVRs because xMega has not improved one bit with age.... and then some guy in QA says "Uh the differential ADC is totally hosed"
 
What do they do? Do they risk slowing down the do-or-die project for the AVR architecture to fix this part? Do they throw the whole thing out entirely? 

None of the above. They tell documentation that the part's are shipping, and to hurry up and remove all mention of the differential ADC from the datasheet because it doesn't work and they're not going to do a silicon rev to fix it. Hence, the datasheet gets hastilly and sloppily edited to remove references to the differential ADC, but anyone reading it can tell it was supposed to be there. 

So the 828 ships as is (I'm certain that the busted ADC is still there, and that someone who had time to spare could work out what was so badly broken about it that they airbrushed it out of the datasheet. I also think it's likely that a small number of chosen customers know the whole story, and that there's a workaround, but one that's too much of a awkward shitshow, or is to embarrassing to  disclose publicly, but it worked to get what those customers needed done, and they are the ones buying and using most of them), and recognizing the part as a humiliating failure, they don't bother to polish the headers or anything, and nobody ever went back and put the PUE bit names into the header. 

The above is 100% speculation and theorizing, and I have proof, no smoking gun - not even a shell casing, just a few bullet holes in the datasheet.... But that's my theory on those last three classic tinies. All of them were very rushed at the end. Look how the 841, the best of the lot, didn't get it's PUE registers put into the IO space! The 1634 had the same bug with the bad pin! They had a team with keyboards with only the Ctrl, C, and V keys on them doing life-extension on the classic ATmegas in the form of the PBs, they did the mega x8PBs, then they started on the x4PBs, released the 324PB, and before they can do the 644PB or 1284PB, those guys get yanked off that project and thrown at some other project (probably the megaAVR 0-series, eg, the 4808 and 4809 and their smaller-memory variants). That's what panic looks like from the outside. And the panic was well founded - the parts weren't ready soon enough to keep Atmel from having to sell out to Microchip right as they launched them (and they launched them with incredibly slapdash testing - the errata on the first modern AVRs released had a table of contents that ran over a page, and it wasn't really until 2022 that they got a modern AVR out the door that didn't have any serious silicon bugs (that would be the DD-series). That's why the peripherals on the modern AVRs look like the engineers found them in the junkyard out back, scavenging xMega and classic AVR for pieces that they can reuse on the modern AVRs - because that's what they were doing!

Again - this is all _slanderous speculation with no evidence to support it_ other than that the publically known history of the AVR architecture over the past dozen+ years is impossible to rationalize any other way. :-P 

And that's why the PUE bit names aren't provided by the headers.

**Comment by fxframes** [comment]
Thanks for the replies!
I can always add an external resistor for the pull-up if I decide to keep the 828. TBH the 828 was being used just because at some point last year it was one of the few readily available, but I might swap it for something a bit more manageable down the road.
Thanks again.

**Reply by SpenceKonde** [reply]
I mean, the pullup registers are 100% accessible, and they work.  And the PUEx names are defined in the headers. And pinMode works. It's just the PUExn names that are missing. 

And PUExn, the defines that aren't there, would simply have a value of n. It's not like they're hiding some confusing or mysterious magic number. 

I've always found it sort of ridiculous that the headers ever provided the PORTxn, PINxn, and DDxn bit names as #defines on classic AVRs. 

I had also noticed that the 828 seemed to be one of the few readily available parts throughout the chip shortage. It's cause nobody wants the damned things! Even before the days of modern AVRs, they had little to recommend them. But in light of all the snazzy modern AVRs coming out with peripherals that blow the classic AVRs out of the water, more favorable pricing, and other improvements, the 828 is has gone from an already disappointing part to a truly dismal one. Like, a 32 pin AVR DD-series is $1.25 with 16k of flash. 7 cents more gets you double the flash and ram with a 32DD32, and 12 cents ontop of that gets you a 64DD32, with 64k of flash, 8k of ram. The DD's have an ADC with 2 more bits of resolution,  on almost every pin, dual serial ports (and that's just the DD's - top end DA/DB parts come with 3 in 32-pin, 5 in 48-pin and 6 in 64 pin versions!) with fractional baud rate generators so you can do 230400 baud or more even at speeds like 8 MHz no problem (the maximum relative to system clock is the same, but you can actually approach it in practice, which you couldn't on classics because of rounding errors), the internal oscillator is more accurate than non-828/1634/841 parts, and doesn't have that crazy dependance on voltage when running above 4V that the 828/1634/841 has, it can run at 24 MHz over the whole 1.8-5.5v voltage range... All that, and it has lower power consumption. And a DAC. And higher max operating temperatures (125C for extended temp range). 

Oh, and if you like to live on the edge, you can overclock the hell out of them at room temperature as a result of the headroom needed to ensure they work at those crazy high temperatures. (40 MHz with crystal and ext temp range is stable on most specimens, and at least some of the E-spec parts can even manage a 2:1 overclock with an external clock, while the PLL that can clock the high speed type D timer generally can be pushed even harder. Most can do a 2.67:1 overclock (32 MHz and 4x multiplier for 128 MHz vs rated 48 MHz) on almost every specimen (again at room temp). While I haven't looked all that hard, I've never had it work when I try to clock TCD from the PLL at 40MHz x 4 = 160 MHz. 

Modern tinyAVRs give you pieces of that functionality, and most of the peripherals are the same, they just have fewer of them and can't be run at the same crazy speeds. The 1-series has a (non-PLL) TCD, a second ADC (in lieu of a differential one - I think the idea is that you'd kick off two conversions at the same time and subtract them to simulate that, which I suspect is what the "almost a differential ADC" on the Dx series is doing under the hood). The 2-series finally has a *true* differential ADC like classic AVRs had, only with higher resolution and more flexibility with pins (that ADC is learning a couple of new tricks while reducing the number of confusing knobs to twiddle to impact accuracy which had proliferated on the 2-series in preparation for the launch of the next "full size" line of AVRs, the EA-series which will feature what is largely an identical ADC), which I expect to see potentially by the end of the year, based on their release timing lately, though it of course depends on how smoothly development is going, 

(sometime around 2020, coincidentally right around the time of a disastrous launch of the AVR32DA parts that involved parts actually getting recalled after they'd shipped to customers because interrupts were completely broken and they'd managed to not notice. They hushed up the whole mess, and the only official confirmation of it you can find is one engineer posting on avrfreaks telling people to contact microchip support for replacements - ever since then they seem to have gotten more willing to let release dates slide a bit in order to fix bugs and release products that aren't as much of a mess, and they've gotten the errata lists on new parts short enough that it'll fit on the screen without scrolling. But even without a dramatic incident like that, they have a lot more latitude to do that now that they have a bunch of modern AVRs for sale. Since the dawn of the modern AVR era, they've launched... I think 70 new AVR products in like 6-7 years, almost a product a month.

For the whole modern AVR era until than (ie, 2016-2020), it sorta felt like they were packing a not-quite-done product up, and throwing it to customers from the back of a moving truck before flooring it back to the lab to work on the next batch of parts, with the music cranked up real loud so they don't hear users asking about fixes for silicon bugs. 

They still don't stop the truck, and the music is still blasting, but the parts they're chucking off the truck work better, and instead of driving the truck down the highway at 95 mph and ignoring red lights on the way, now they're only doing 80 mph and just blowing past stop signs, and the ongoing capacity shortage is such that there's a mob of customers fighting over every tray of parts they throw and a 1 year backorder queue. Which has led to this weird dynamic where you can get the parts released yesterday, but the parts released last month have sold out and are already backordered into next year. 

It's eased a bit since the worst of it, ATtiny3217's were trading for as much as $5/ea from scalpers in 2021. They were released in i think 2017 or 2018, were available easily for a couple of years, but they were one of the most severely impacted parts when the chip shortage hit. I think Microchip has two bottlenecks - like everyone, they don't have enough fab capacity, but they also don't seem able to turn as much of their production into QFN packages as buyers want. If that wasn't a bottleneck, you'd expect that it would be equally hard to find all package variations - but in fact, it's consistently harder to get QFNs compared to anything else. I was able to score enough 1626s in QFN thanks in part to 50-per-person-limit deal a few months ago) and 3227's because I got a bunch the instant they were released, but the 3226's I only have in SOIC. The SOIC 3226 occasionally appears in stock somewhere but the QFN version is what I'd really like to get my claws on. You can fit that between the holes on a DIP-20-sized PCB to convert it to something that fits into a DIP-20 socket (I'm going to launch a product consisting of a modern AVR on a DIP-adapter like that, for the ATtiny 1616, 3217, 3227, 1626, 1624 and 412 in the coming days, for SMD averse hobbyists, but were it not for shortages, I'd be launching with 3226 and 3224s instead of 1626's and 1624's. AVR32DD20 boards like that are coming too, but it requires some tweaks to the PCB). DIP-20 is still smaller than that massive wide SOIC-20 thing. 

Wide SOIC-20, I think deserves the title of "crappiest package still common on production parts" (that category doesn't count the difficulty of hand soldering - "hardest non-BGA package to solder at home" is a fight between a number of parts with 8-16 pins in modified QFN packages with no exposed pad - uQFN-10, uQFN-8, and uQFN-16 are the ones I'm thinking of (BGAs are a different category, as they're unspeakably hard to deal with at home, regardless of the tools you have, in no small part because you only get one shot. If you have to dismount the chip, it needs to be reballed before you can try again). All the QFN and DFN parts are kind of a bitch to hand solder, but the ones with no exposed pad are hard to solder even in the reflow oven, because there's no exposed pad to make them self-center like there is on a normal QFN. That self-centering is a bigger deal than one who doesn't have experience with reflow/hotplate/hot-air soldering would ever imagine. `*` Those things are bigger than the side of a house. The next size down, SOIC-16, for just 4 fewer pins, is less than 1/3 the thickness and less than half as wide. I think a narrow SOIC-20 package made by just extending a SOIC-16 to 20 pins, would be too long and thin, and would be too prone to breaking from board flex, and nobody ever did an intermediate width SOIC for 18-24 pins. Really, JEDEC (I think that's the relevant standards group) could have and should have made the wide SOIC a little less wide, and a little thinner, since they only have to get up to 28 pins - the next size up is 32 pins, and nobody even wants a 32-pin SOIC or SSOP - that's QFP territory (though these days, 20+ pins to many designers is QFN territory, that's a more recent change). The dimensional scaling on the (T)SSOP packages is a lot more reasonable than the SOICs. I think those widen at 24 or 28 pins, but even so not nearly as dramatically. FFS, you can fit a 32-pin QFP into less space than a 20 pin wide SOIC, by nearly a factor of two. 

`*` I get slightly higher yield-before-rework, and much higher yield-without-hotplate/hotair rework on a QFN-48 even with 0.4mm pin pitch than I do on a TQFP-48 with 0.5mm pitch, though QFP-32 @ 0.8mm pitch is easier than QFN32 @ 0.5mm pitch, largely because my yield-without-hotplate/hotair is apx. 100% for the TQFP-32. (Its 50-70% for QFP-48 and and around 60-80% for... really most QFNs (the QFN-48 needs more precise alignment because of the finer pitch and the fact that less accidental rotation is needed to give problems that need hotair/plate to fix, but the larger exposed pad compensates with self-centering),  and without any rework, those yields are around 80% for the QFP-32 around 30% and something like 40-60%, respectively. Yields for 64-pin QFPs are about the same. I haven't done 64-pin QFNs yet; on the AVRs 64-pin QFN is 0.45mm pitch or something). There is a very high scatter for the QFP, especially QFP32, due to extreme dependance on the solder paste and the job I did stenciling it - non-rework yield has ranged from 0% using a brand of solder which in hindsight is clearly unfit for purpose (not even unleaded! 63:37 eutectic tin lead, just an extraordinarily poor formulation from a brand that I've since determined is rarely even of average quality and never above average, despite the breadth of their offerings), up as high as 100% when I nail the solder paste application and used good paste), as those strongly impact the number of solder bridges, which on QFP-32, are super easy to fix), 

Those are just guesses at the yields that **I** get in **my** soldering process. The influence of the solder paste brand and whether I did a good job on the paste application or a lousy one is a significant confounding factor, particularly since I've been experimenting to try to find a lead-free paste that works in my reflow oven, which can't *quiiiite* do 217C SAC305 even after applying the usual modifications (there are modkits that should fix that, but the pricing is well over $100, and involves replacing almost every component connected to a wire except for the heating elements). Pro tip: all leadfree solder that melts below 200C is either lying about being lead free, is absolutely awful solder that is completely unfit for purpose, or else it's the 138C SnBi eutectic, which has a number of problems (brittleness, garbage thermal conductivity, excessively low melting point, inferior wetting, and harder rework), with only two specific exceptions. And I'm only confident that 1 of them is lead free `**` Regardless of the package, my yields are >95% after hotplate/air for alignment issues where needed followed by debridging with iron and braid, for all the packages I've used (miniQFN16 may be an exception, but the sample size is too small. SOIC is 95-100% with no rework at all, barring paste mishaps, and for 2-terminal passives down to 0603 and SOT-23/SOT-23-6, it's even higher, and the rework for any of those is trivial when it is needed, and all of the SOIC-or-easier ones except 0603 are easypeasy to solder by hand. MSOP-10 on the other hand is quite hard via reflow or by hand - and surprisingly, micro USB ports are also pretty bad, though much easier to rework. 


`**` When I launch my Patreon, one thing patrons will get is access to is my treatise on soldering options which is currently 180kb in size as a plain markdown formatted text file, with tens of thousands of words (pasting it into word shows 31400 words currently, but I think it's counting numbers in tables and possibly formatting markup as words, because I know the words I type there average >6 letters each) and is based on a lot of empirical research with solder that is actually available for purchase and tools that are not unreasonable for hobbyists and small businesses to have. It's cost me several hundred dollars (buying many many brands of solder paste, mostly), has taken easily over 100 hours of work, including writing, experimental work and a wide-ranging study of available literature, covering practical soldering considerations, product ratings (including the identity of that solder with the unique alloy that melts under 200C and is both lead free and good enough to use in place of tin-lead solder) as well as a great deal of historical context, background, and some discussion of adjacent subject matter ranging from metallurgy, the AliExpress economy (including what soldering supplies you should just buy on aliexpress, and which you should buy from reputable western suppliers), and heavy metal toxicity, including historically interesting anecdotes, like the death (which he absolutely deserved!) of a sickeningly unscrupulous funeral home worker that serves as a demonstration of a factor other than raw toxicity that makes several heavy metals particularly dangerous).
